rlpx/handshake.go

```diff
@@ -0,0 +1,138 @@
+package rlpx
+
+import (
+	"crypto/rand"
+	"errors"
+	mrand "math/rand"
+	"net"
+
+	"github.com/decred/dcrd/dcrec/secp256k1/v4"
+
+	"github.com/indexsupply/x/bint"
+	"github.com/indexsupply/x/ecies"
+	"github.com/indexsupply/x/enr"
+	"github.com/indexsupply/x/isxsecp256k1"
+	"github.com/indexsupply/x/rlp"
+)
+
+const (
+	authVersion  = 4
+	prefixLength = 2
+)
+
+type handshake struct {
+	isInitiator     bool
+	localPrvKey     *secp256k1.PrivateKey
+	localEphPrvKey  *secp256k1.PrivateKey
+	remotePubKey    *secp256k1.PublicKey
+	remoteEphPubKey *secp256k1.PublicKey
+	initNonce       []byte
+	receiverNonce   []byte
+}
+
+func newHandshake(localPrvKey *secp256k1.PrivateKey, to *enr.Record) (*handshake, error) {
+	h := &handshake{
+		remotePubKey: to.PublicKey,
+		localPrvKey:  localPrvKey,
+	}
+
+	return h, nil
+}
+
+func (h *handshake) createAuthMsg() (rlp.Item, error) {
+	var err error
+	// initialize random nonce
+	if h.initNonce == nil {
+		h.initNonce = make([]byte, 32)
+		_, err := rand.Read(h.initNonce[:])
+		if err != nil {
+			return rlp.Bytes(nil), err
+		}
+	}
+	// Generate ephemeral ECDH key
+	if h.localEphPrvKey == nil {
+		h.localEphPrvKey, err = secp256k1.GeneratePrivateKey()
+		if err != nil {
+			return rlp.Bytes(nil), err
+		}
+	}
+	// Create shared secret using ephemeral key and remote pub key
+	var sharedSecretBytes [32]byte
+	copy(sharedSecretBytes[:], secp256k1.GenerateSharedSecret(h.localPrvKey, h.remotePubKey))
+	var signedPayload [32]byte
+	for i := 0; i < 32; i++ {
+		signedPayload[i] = sharedSecretBytes[i] ^ h.initNonce[i]
+	}
+	// Sig = Sign(Ephemeral Private Key, Shared Secret ^ Nonce)
+	sig, err := isxsecp256k1.Sign(h.localEphPrvKey, signedPayload)
+	if err != nil {
+		return rlp.Bytes(nil), err
+	}
+
+	rawPubKey := isxsecp256k1.Encode(h.localPrvKey.PubKey())
+	// Auth Body = [Sig, Raw Pub Key, Nonce, Version]
+	return rlp.List(
+		rlp.Bytes(sig[:]),
+		rlp.Bytes(rawPubKey[:]),
```

Comment:
Perhaps we should have a Secp256k1PublicKey function to match the item method.

Reply:
Yeah good idea, I did see that method added in your PR